ethcore/src/miner/transaction_queue.rs

```diff
@@ -290,13 +296,49 @@ impl TransactionSet {
 	fn clear(&mut self) {
 		self.by_priority.clear();
 		self.by_address.clear();
+		self.by_gas_price.clear();
 	}
 
 	/// Sets new limit for number of transactions in this `TransactionSet`.
 	/// Note the limit is not applied (no transactions are removed) by calling this method.
 	fn set_limit(&mut self, limit: usize) {
 		self.limit = limit;
 	}
+
+	/// Get the minimum gas price that we can accept into this queue that wouldn't cause the transaction to
+	/// immediately be dropped. 0 if the queue isn't at capacity; 1 plus the lowest if it is. 
+	fn gas_price_entry_limit(&self) -> U256 {
+		match self.by_gas_price.keys().next() {
+			Some(k) if self.by_priority.len() >= self.limit => *k + 1.into(),
+			_ => U256::default(),
+		}
+	}
+
+	/// Insert an item into a BTreeMap/HashSet "multimap".
+	fn insert_item(into: &mut BTreeMap<U256, HashSet<H256>>, gas_price: U256, hash: H256) -> bool {
+		into.entry(gas_price).or_insert_with(Default::default).insert(hash)
+	}
+
+	/// Remove an item from a BTreeMap/HashSet "multimap".
+	/// Returns true if the item was removed successfully.
+	fn remove_item(from: &mut BTreeMap<U256, HashSet<H256>>, gas_price: &U256, hash: &H256) -> bool {
+		if let Some(mut hashes) = from.get_mut(gas_price) {
+			let only_one_left = hashes.len() == 1;
+			if !only_one_left {
+				// Operation may be ok: only if hash is in gas-price's Set.
+				return hashes.remove(hash);
+			}
+			if hashes.iter().next().unwrap() != hash {
+				// Operation failed: hash not the single item in gas-price's Set.
+				return false;
+			}
+		} else {
+			// Operation failed: gas-price not found in Map. 
+			return false;
+		}
+		// Operation maybe ok: only if hash not found in gas-price Set.
+		from.remove(gas_price).is_some()
```

Comment:
would like to, but `hashes` has captured an `&mut` of `into` in this scope so it has to escape before it can be `remove`d.